src/components/annotations/draw.js

```diff
@@ -73,6 +73,26 @@ function drawOne(gd, index) {
     drawRaw(gd, options, index, false, xa, ya);
 }
 
+// Convert pixels to the coordinates relevant for the axis referred to. For
+// example, for paper it would convert to a value normalized by the dimension of
+// the plot.
+// axDomainRef: if true and axa defined, draws relative to axis domain,
+// otherwise draws relative to data (if axa defined) or paper (if not).
+function shiftPosition(axa, optAx, dAx, gsDim, vertical, axDomainRef) {
+    if(axa) {
+        if(axDomainRef) {
+            // here optAx normalized to length of axis (e.g., normally in range
+            // 0 to 1). But dAx is in pixels. So we normalize dAx to length of
+            // axis before doing the math.
+            return optAx + (vertical ? -dAx : dAx) / axa._length;
+        } else {
+            return axa.p2r(axa.r2p(optAx) + dAx);
+        }
+    } else {
+        return optAx + (vertical ? -dAx : dAx) / gsDim;
+    }
```

Comment:
To improve performance, we may consider refactoring:
```js
var delta = vertical ? -dAx : dAx;
return !axa ? optAx + delta / gsDim :
    axDomainRef ?
        optAx + delta / axa._length : // here optAx normalized to length of axis (e.g., normally in range 0 to 1). But dAx is in pixels. So we normalize dAx to length of axis before doing the math.
        axa.p2r(axa.r2p(optAx) + dAx);
```

Reply:
I think inline-ifs are good if they make the code clearer, but here I would not say they do that haha :-)